### PR TITLE
boot time improvements

### PR DESCRIPTION
**Pull request by arthurnn**
### Problem
Rails is slow. After a few months of development, my Rails app takes more time to load, it is not as fast as it used to be. 😢 

Note: I know this is not a rubygems.org problem per-say, however this is a really common problem in Rails apps, and usually devs dont know what to do. And also, this isn't a Rails issues. So I think someone needs to start given a good example, and hopefully the community will start following some good patterns and have apps that boot faster.

I decided to take a look, and see how far I could get and to try to cut down boot time as much as possible.

### Main idea
- Remove `Bundler.require` code
- Lazy load as much as possible
- Cache paths on require, to speedup rubugems require code

Besides speed there are other advantages to lazy load things:
- Things are a bit less magical, and we can see better what we are using
- Future gems added to Gemfile, wont slow things down

Similar work done by @indirect , and posted here http://andre.arko.net/2014/06/27/rails-in-05-seconds/ 👏 

thoughts @evanphx @qrush @dwradcliffe @sferik 
cc @rafaelfranca @schneems

#### Show me some numbers
Before:
```
[arthurnn@snowball rubygems.org]$ time bin/rails runner '0'

real	0m3.214s
user	0m2.516s
sys	0m0.671s
```

After:
```
[arthurnn@snowball rubygems.org]$ time bin/rails runner '0'

real	0m1.778s
user	0m1.591s
sys	0m0.177s
```

### CODE REVIEW

**Comment by qrush** [comment]
Is _one second_ of Rails boot time worth the long-term maintenance headache of not using vanilla/community standard `Bundler.require`? I'm not sure if this is really worth it.

**Comment by schneems** [comment]
This could have bad memory impact if each fork is lazily loading the same large libraries and not taking advantage of CoW.

**Comment by indirect** [comment]
@schneems I'm assuming that in production mode, the prefork block will call `Bundler.require(:production)` or whatever, so that shouldn't be a problem.

@qrush napkin math: for 10 devs running 25 `rails c`, `rails s` or `rake ` commands per day this saves 40 hours per year (given the benchmarked 1.5 second speedup). That makes it sound like a good deal even if it adds 10 hours a year in extra maintenance costs.

I'm personally a fan of this, but I also wrote the blog post cited in the PR :stuck_out_tongue_closed_eyes:

File: Gemfile [hunk 1]
```diff
@@ -2,23 +2,25 @@ source 'https://rubygems.org'
...
 gem 'rails', '~> 4.2.5'
 gem 'rails-i18n'
 
 gem 'autoprefixer-rails'
 gem 'aws-sdk-core'
-gem 'builder'
+gem 'bootscale'
 gem 'clearance'
 gem 'clearance-deprecated_password_strategies'
+## TODO maybe remove deamons?
```
Comment (simi) [inline Gemfile]: :scissors: s/deamons/daemons/g

**Comment by simi** [comment]
This is technically great work, but I think this is not the "rails way" and this can lead to lot of confusions and to lot of questions.

* Do you still need Bundler to handle dependencies since rubygems (experimentaly?) can install gems from Gemfile now? 
* Does this mean `Bundler.require` is slow and you should not use it?
* Is the 1.5s boot improvement really worth it and perceptible for developers? *@indirect I think your example is only theoretical math not proving any real benefit.*
* If you think dev environment is slow, why no preloader is used? (I'm not using preloaders also, but I think this is common pattern in Rails how to solve this problem.)
* What should be autoloaded and what shouldn't be? (I think only external files are not autoloaded in this change.)

:question:

Actually I'm also big fan of this, but some questions need to be answered and this change needs to be **documented** in the README.md and CONTRIBUTING.md files since with this change this project is leaving the "rails way".

tumblr_m7j16zqvso1rbl0zpo1_400

File: script/delayed_job [hunk 2]
```diff
@@ -1,5 +1,12 @@
...
 
 require File.expand_path(File.join(File.dirname(__FILE__), '..', 'config', 'environment'))
+
+require 'delayed_job_active_record'
 require 'delayed/command'
+
+Delayed::Worker.max_attempts = 3
+Delayed::Worker.max_run_time = 5.minutes
+Delayed::Worker.logger = Logger.new(Rails.root.join('log/delayed_job.log'))
+
```
Comment (dwradcliffe) [inline script/delayed_job]: We don't use this script in production so this won't work. We use a rake task to run job workers.

**Comment by qrush** [comment]
> This is technically great work, but I think this is not the "rails way" and this can lead to lot of confusions and to lot of questions.

Agreed on this. We're diverting from the "golden path". Is *one second* of boot time actually worth that? Over the years that RubyGems.org will need to stay maintained and running for the community? If it was an order of magnitude change (say, from 10 seconds to 1 second) that would be an easy and obvious win.

I think we need to think long-term here. I'm hoping that this Rails app will stay alive as long as Ruby does. Or until something better comes along. Given a timespan of _years_ to _decades_ - is diverting from the community-accepted path of using + requiring gems (what this application helps do!) worth it? I think it's questionable.

Here's a concrete example: Are these changes going to make RubyGems.org easier to upgrade to Rails 5? I doubt it.

**Comment by arthurnn** [comment]
> This could have bad memory impact if each fork is lazily loading the same large libraries and not taking advantage of CoW.

@schneems This should not change anything regarding CoW, as in production we still would eager load the app folder which would require everything before forking the process.

@qrush @simi 
Yes this diverted a bit from the "golden path". I wanted to do this experiment to see what other people think about it, and how far are we going outside the rails way. Maybe, this isn't going too far. The only thing different here is, you need to require what you use.

> Is one second of Rails boot time worth the long-term maintenance headache...

I think this is not about the one second cut. This is more about the long term. A common pattern is that rails app boot time starts small and linearly grow as you add Gems. Thats the big win here. One example was rubocop, that by itself was taking about 400ms on every boot load. (even if we dont use it in runtime)

> Here's a concrete example: Are these changes going to make RubyGems.org easier to upgrade to Rails 5? I doubt it.

I am pretty positive this wont change anything on the upgrade process.

>  this change needs to be documented

For sure, yes! But I think if you know ruby, this should be obvious, we maybe just need to say we remove the auto-require magic bit.


What this will change most is the ADD GEM process, which IMO should not be as easy as a drop a gem to the Gemfile. Maybe, because of this easy process, we have a lot of rails apps that are slow to boot, and we need to hack something like Spring/Zeus to make them faster.


FWIW: I am a big fan of the Rails way, and I try to stick with it in 95% of the time. However I would like to bring up this debate on propose, so we can maybe re-think the rails way of auto-requiring gems and see if this is an actually good solution or not.

Maybe people from the rails team could chime-in too. cc @dhh @tenderlove @fxn

**Comment by nateberkopec** [comment]
> One example was rubocop, that by itself was taking about 400ms on every boot load.

jesus, can we get this one as a separate PR? I think that one's worth it on it's own.

**Comment by dhh** [comment]
This is exactly the problem Spring was designed to deal with. I remain convinced that it's folly to treat the local Rails instance as it was a CGI process that needs to be rebuilt and thrown away on every run. Spring treats the development process like FCGI, so repeated test runs, command runs, and the like do not have to evaluate the whole app (and all associated frameworks from scratch).

**Comment by dhh** [comment]
All the latest Spring benchmarks are available here: https://github.com/rails/spring – they're far better than this, imo, and much less work to maintain. And The Rails Way.

**Comment by schneems** [comment]
Spring is awful if you're working on libraries like Rails, puma, rack, really any gem. It may save you miliseconds of boot a few hundred times a week, but it will cost you a few hours/days/weeks of bug hunting before you realize that all those changes to libraries aren't actually being used by your local system. 

I have:

```sh
export DISABLE_SPRING="true"
```

In my bashrc. Even with spring, cold boot time still matters. When you deploy to production, you're in a fresh process, maybe you're developing on multiple apps using docker containers locally that are spun up and down. Or you're debugging example apps (like the kind that come in attached to issues) that you'll only ever need to boot once. If you need to boot again, you have to restart spring due to modified dependencies. 

Spring is a bandaid for slow booting apps, not a fix.

**Comment by qrush** [comment]
I'd much rather go with Spring than deal with the long-term maintenance headache of having `require`s littered all over the app.

I'd once again want to question: Is an app that takes 3 seconds to boot a "slow booting app"? There's bigger fish to fry here, IMO.

**Comment by dhh** [comment]
Richard, we clearly differ on our assessment of Spring. Incredibly little
time is spent by most developers working on Rails/puma/any infrastructure
gem concurrently with their app development. And for those rare, few cases,
you can indeed just disable Spring. But for all the other time, which is
nearly all of it, EVEN in my case, and I work a lot on Rails, Spring is a
fantastic step forward.

Yes, working on boot time is nice, but as Nick says, littering your code
with requires and other hoops to do so is not. No way I'd litter my code
like that to get a initial boot speed-up that's going to be felt hardly any
of the time I'm doing actual work.

Spring is in my opinion the overwhelmingly obvious choice for fixing the
PROGRAMMER HURT associated with large apps.

**Comment by rafaelfranca** [comment]
Have we measured what is slow? Changing `bundle.require` for `require` calls is technically the same thing, unless `Bundler.require` does more that requiring gems, and if that is the case I think it would be better for the community as whole if we fix that.

**Comment by indirect** [comment]
@rafaelfranca the slowness is that `Bundler.require` forces eager loading gems in development, when it is not needed. Making gems require faster is good, but cannot ever be as fast as not requiring those gems until they are needed.

@qrush @dhh my experience with Spring is apparently as bad as Richard's, and I almost never actively work on Rails or Rack. I have tried both using spring and using explicit require in a Rails app to speed up booting, and Spring was by far the more frustrating option with more debugging time spent on it.

Honestly I'm a little surprised that this is so contentious... require statements at the top of files that use those libraries just seem like plain good Ruby practice to me. Adding a require statement to the file that uses a gem after you add that gem to your Gemfile also seems pretty non-controversial to me. Using `Bundler.require` to slowly load every single gem into your Ruby interpreter when all you want is to run a single unit test has always struck me as frustrating and pointless.

Is it possible that we can all agree that clear requires are good, and then let spring users have fast boot via spring and let non-spring users have fast boot times by avoiding `Bundler.require`?

**Comment by dhh** [comment]
André, it would be helpful if you, and others, who hit trouble with Spring
help improve matters by filing bug reports. I've been using Spring for the
last two years developing Basecamp 3 and it's been a massive boost. Yes,
there is the rare case where you have to "spring restart", but weighed
against the benefits, it has seemed a small sacrifice to me. Jon is
actively developing Spring, so maybe it's time to give it another try?

I do not consider require statements at the top of files a best practice in
Rails applications. This is why Rails does auto-loading of models,
controllers, helpers, and more. To avoid having to specifically spell out
requires. I consider the same benefit to hold for libraries that have
already been mentioned in the Gemfile.

**Comment by schneems** [comment]
I'm with Nick on the overall change, this is a little too off the beaten path. However I think there are still nuggets of speed we can safely apply, maybe moving rubocop, or opening an issue with their repo to see why it is taking so long to load.

We have differing opinions of the value of spring and that's fine, especially since I can opt out globally. While spring may alleviate hurt, it doesn't take away the cause. When we can take reasonable actions to make things faster we should. I'm in favor of benchmarking and exposing these types of issues. Thanks for the work @arthurnn, definitely some good stuff to think about.

**Comment by mediafinger** [comment]
While I like the idea of spring, I do agree with @schneems and @indirect: after having spent twice an hour (or more) of bug hunting because not running `spring stop && spring start` those saved seconds seem to be of no avail compared to the possible costs.

The increased maintenance costs of explicit `require`s mentioned by @qrush I don't see. Developers might need to understand which gem or module defines which methods and behaviors better than now.
On the other hand, having the `require`s explicitly in the classes where the required modules are used could lead to easier understanding of code for developers new to the project. And it could make the refactoring costs for replacing a gem clearer.

But yes, explicit `require`s are more to type, not that nice to the eye and not The Rails Way. Explicitness and what people came to know as Rails Magic seem to exclude themselves.

@arthurnn: thanks for demonstrating this way of speeding up boot time.

**Comment by jvanbaarsen** [comment]
I honestly don't see the problem with Spring, I think I have about 10 apps that I'm actively developing on, and non of them are causing issues. I think I had to do `spring restart` once. Having `requires` everywhere makes it hard for people to jump on to the project since it is not the Rails way.

@mediafinger Did you file a bug with spring when you ran into this issues? Since I feel those need to be solved instead of worked around.

**Comment by mediafinger** [comment]
@jvanbaarsen I can not remember what caused my first issue with spring. The second time I was extracting code from our app into a gem - and in the Gemfile I required the code from a local repo. It just slipped my mind some day that changes in this other repo won't get autoloaded. I assume this is expected behaviour and not a bug in spring.

**Comment by ktheory** [comment]
Really neat discussion. :smile:

Just wanted to highlight @indirect's comment:

> I'm assuming that in production mode, the prefork block will call `Bundler.require(:production)`

That's great for *deployment operability* as well as CoW memory.

If a syntax error or uninitialized constant error somehow gets deployed, it'll fail fast and can be rolled back. Otherwise with lazy-loaded code, a deployment could seem successful, only to fail when a particular file is loaded.

**Comment by qrush** [comment]
Fantastic point @ktheory.

I think we should lay this one to rest. We can get new PRs fired up for dealing with rubocop and other large gems. How's that sound @arthurnn ?

**Comment by arthurnn** [comment]
> If a syntax error or uninitialized constant error somehow gets deployed, it'll fail fast and can be rolled back. Otherwise with lazy-loaded code, a deployment could seem successful, only to fail when a particular file is loaded.

Yep, but in production code is not lazy-loaded. With the eager load flag on everything will be preloaded. So I dont understand how would that be a problem.

**Comment by arthurnn** [comment]
@qrush I think this was a productive discussion. Will close it, and work in a few follow up points. However will keep the "Rails Way" , and invest my time on spring.

Thanks everyone. ❤️

**Pull request closed without merging**